Entity Type Registry

halotroop2288 opened the pull request:
A simple invoker for `EntityList.addMapping` with an ID tracker separate from the block and item registry.

I think this should be moved to a separate registry class in future versions, with `GameRegistry` being renamed to `BlockAndItemRegistry` (or something like that). Maybe even separate those two into separate classes. Otherwise, it's bound to get extremely bloated.

halotroop2288 committed:
Entity type registry method

Fox2Code commented:
Mappings are not implemented, this is not a hard requirement, but would be nice.
(Making server send mappings to client)

Also did you tested that with a client connecting to a server?
(With a mod that register a custom entity)

I assume singleplayer is fine.

halotroop2288 committed:
Entity type id mapping

Fox2Code commented on common/src/main/java/com/fox2code/foxloader/loader/packet/ServerHello.java:
```diff
@@ -9,64 +10,92 @@
                 // Next field is how much backward compatible is the packet
-                dataInputStream.readUnsignedShort() > SERVER_HELLO_VERSION_NEXT) {
+                inStream.readUnsignedShort() > SERVER_HELLO_VERSION_NEXT) {
             throw new RuntimeException("Client is critically out of date, please update FoxLoader");
```
Keep this code in mind for later

halotroop2288 replied:
This part is hard to read because of the broken English unfortunately. I'm not sure I understand what `SERVER_HELLO_VERSION_NEXT` is actually for.

I was expecting to find something like this:

```diff
public class ModLoader {
	/**
	 * V0 reads item/block registry data
	 * V1 reads item/block registry metadata
+	 * V2 reads entity type registry data
	 */
-	public static final int SERVER_HELLO_VERSION = 1;
+	public static final int SERVER_HELLO_VERSION = 2;
	/**
	 * The version of ServerHello that the client can read without issue
	 * This version must be less than or equal to SERVER_HELLO_VERSION
	 */
	public static final int SERVER_HELLO_BACKWARD_COMPATIBLE_VERSION = 0;
```

```java
	public void readData(DataInputStream inStream) throws IOException {
		final int serverHelloVersion = inStream.readUnsignedShort();
		if (serverHelloVersion <= SERVER_HELLO_BACKWARD_COMPATIBLE_VERSION) {
			// Client disconnects here because it can't read the packet.
			throw new IOException("Client is critically out of date. Please update FoxLoader.");
		}

		// ...
	}
```

But I don't know how to handle this code if it works differently.

Fox2Code replied:
We can't retroactively change older client code

halotroop2288 replied:
No shit. Got any helpful advice?

Fox2Code replied:
@halotroop2288 you are doing well!

Fox2Code commented on common/src/main/java/com/fox2code/foxloader/loader/packet/ServerHello.java:
```diff
@@ -9,64 +10,92 @@
         }
-        if (dataInputStream.available() == 0) {
+
+        if (inStream.available() < 2) {
```
Maybe a version check here too, this code look fine too

Fox2Code commented on common/src/main/java/com/fox2code/foxloader/loader/packet/ServerHello.java:
```diff
@@ -9,64 +10,92 @@
-        dataOutputStream.writeShort(SERVER_HELLO_VERSION);
-        dataOutputStream.writeShort(this.registryEntries.size());
+    public void writeData(DataOutputStream outStream) throws IOException {
+        outStream.writeShort(SERVER_HELLO_VERSION);
```
We should write what we except minimum packet version is supported for reading (In this case `0`)

Maybe adding a `public static final int  SERVER_HELLO_MIN_SUPPORT = 0;` ?

halotroop2288 replied:
I tried to make the changes to ServerHello minimally invasive. It should be backwards compatible the way it is, as the entity type data is written to the end and it only reads the next data if it exists.

Fox2Code replied:
It isn't for now because of how I made the packet, the second version is ti tell how backward compatible the packet is.

halotroop2288 replied:
Why would you need to send that information, except to tell the server which version of ClientHello it accepts?
The ServerHello version only needs to be checked for compatibility on the client and vice versa.

Fox2Code replied:
Older FoxLoader client will except ClientHello to send 2 short if version is `1` or higher.

We cannot change older client code unless you have a time machine.

halotroop2288 replied:
??? There's nothing that needs to be changed on the older client.
They'll just stop reading at the end of the block and item metadata.

Fox2Code reviewed (changes requested):
The code you did show a deep understanding of FoxLoader

Fox2Code commented:
The DataInputStream is backed by a ByteArrayInputStream, so not reading the packet fully is fine.

This means if we add data at the end of the packet, old versions can safely ignore it.

halotroop2288 commented:
^ My thoughts exactly.

halotroop2288 committed:
[WIP] Prevent older clients from connecting

Prevents older clients from connecting, which would cause desync issues.
My mind is exhausted from trying to figure out how this should work, so I'm just pushing it as-is.

Fox2Code commented on common/src/main/java/com/fox2code/foxloader/loader/packet/ServerHello.java:
```diff
@@ -30,13 +41,15 @@ public ServerHello(HashMap<String, RegistryEntry> registryEntries,
-                // Next field is how much backward compatible is the packet
-                inStream.readUnsignedShort() > SERVER_HELLO_VERSION_NEXT) {
-            throw new RuntimeException("Client is critically out of date, please update FoxLoader");
+        int clientBackwardCompatibilityVersion = inStream.readUnsignedShort();
```
If `serverHelloVersion < 0` do not call `readUnsignedShort`, and assume packet version is `0`

Fox2Code reviewed (changes requested):
You did good

Fox2Code commented:
Thanks for everything, I'll take it from there
